libs/datamodel/core/src/transform/ast_to_dml/db/relations.rs

```diff
@@ -0,0 +1,270 @@
+        .db
+        .iter_model_relation_fields(relation_field.referenced_model)
+        // Only considers relations between the same models
+        .filter(|(_, opposite_relation_field)| opposite_relation_field.referenced_model == *model_id)
+        // Filter out the field itself, in case of self-relations
+        .filter(|(opposite_field_id, _)| !is_self_relation || opposite_field_id != field_id)
+        .find(|(_, opp)| opp.name == relation_field.name)
+        .map(|(opp_field_id, opp)| (opp_field_id, &opposite_model[opp_field_id], opp));
+
+    RelationEvidence {
+        ast_model,
+        model_id: *model_id,
+        ast_field,
+        field_id: *field_id,
+        relation_field,
+        opposite_model,
+        is_self_relation,
+        opposite_relation_field,
+    }
+}
+
+fn ingest_relation<'ast, 'db>(
+    evidence: RelationEvidence<'ast, 'db>,
+    relations: &mut Relations<'ast>,
+    ctx: &'db Context<'ast>,
+) {
+    let relation_type = match (evidence.ast_field.arity, evidence.opposite_relation_field) {
+        // m:n
+        (ast::FieldArity::List, Some((opp_field_id, opp_field, _))) if opp_field.arity.is_list() => {
+            // This is an implicit many-to-many relation.
+
+            // We will meet the relation twice when we walk over all relation
+            // fields, so we only instantiate it when the relation field is that
+            // of model A, and the opposite is model B.
+            if evidence.ast_model.name.name > evidence.opposite_model.name.name {
+                return;
+            }
+
+            // For self-relations, the ordering logic is different: model A
+            // and model B are the same. The _first field in source text
+            // order_ is field A, the second field is field B.
+            if evidence.is_self_relation && evidence.ast_field.name.name > opp_field.name.name {
+                return;
+            }
+
+            RelationType::ImplicitManyToMany {
+                field_a: evidence.field_id,
+                field_b: opp_field_id,
+            }
+        }
+
+        // 1:1
+        (ast::FieldArity::Required, Some((opp_field_id, opp_field, _))) if opp_field.arity.is_optional() => {
+            // This is a required 1:1 relation, and we are on the required side.
+            RelationType::OneToOne(OneToOneRelationFields::Both(evidence.field_id, opp_field_id))
+        }
+        (ast::FieldArity::Optional, Some((opp_field_id, opp_field, _))) if opp_field.arity.is_optional() => {
+            // This is a required 1:1 relation, and we are on the virtual side. Skip.
+            return;
+        }
+        (ast::FieldArity::Required, Some((_, opp_field, _))) if opp_field.arity.is_required() => {
+            // This is a 1:1 relation that is required on both sides. Error.
+            return; // TODO: error
+        }
+        (ast::FieldArity::Required, Some((opp_field_id, opp_field, opp_relation_field)))
```

Comment:
You say later in the block that both sides are optional, should this be:
```suggestion
        (ast::FieldArity::Optional, Some((opp_field_id, opp_field, opp_relation_field)))
```